ggshield/oauth.py

```diff
@@ -0,0 +1,274 @@
+    def __init__(self, config: Config, instance: str) -> None:
+        self.config = config
+        self.instance = instance
+        self._oauth_client = WebApplicationClient(CLIENT_ID)
+        self._access_token: Optional[str] = None
+
+        self.server: Optional[HTTPServer] = None
+        self._handler_wrapper = RequestHandlerWrapper(oauth_client=self)
+        self._generate_pkce_pair()
+
+    def oauth_process(self) -> None:
+        """
+        Handle the whole oauth process which includes
+        - opening the user's webbrowser to GitGuardian login page
+        - open a server and wait for the callback processing
+        """
+        # enable redirection to http://localhost
+        os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = str(True)
+
+        # TODO: get the token name from init params
+        self._token_name = "ggshield token " + datetime.today().strftime("%Y-%m-%d")
+
+        self._prepare_server()
+        self._redirect_to_login()
+        self._wait_for_code()
+        click.echo(f"Created Personal Access Token {self._token_name}")
+
+    def process_callback(self, callback_url: str) -> None:
+        """
+        This function runs within the request handler do_GET method.
+        It takes the url of the callback request as argument and does
+        - Extract the authorization code
+        - Exchange the code against an access token with GitGuardian's api
+        - Validate the new token against GitGuardian's api
+        - Save the token in configuration
+        Any error during this process will raise a OAuthError
+        """
+        authorization_code = self._get_code(callback_url)
+        self._claim_token(authorization_code)
+        token_data = self._validate_access_token()
+        self._save_token(token_data)
+
+    def _generate_pkce_pair(self) -> None:
+        """
+        Generate a code verifier (random string) and its sha encoded version to be used
+        for the pkce checking process
+        """
+        self.code_verifier = token_urlsafe(96)[:128]
+        self.code_challenge = (
+            urlsafe_b64encode(sha256(self.code_verifier.encode()).digest())
+            .decode()
+            .rstrip("=")
+        )
+
+    def _redirect_to_login(self) -> None:
+        """
+        Open the user's browser to the GitGuardian ggshield authentication page
+        """
+        static_params = {
+            "auth_mode": "ggshield_login",
+            "utm_source": "cli",
+            "utm_medium": "login",
+            "utm_campaign": "ggshield",
+        }
+        request_uri = self._oauth_client.prepare_request_uri(
+            uri=urlparse.urljoin(self.dashboard_url, "auth/login"),
+            redirect_uri=self.redirect_uri,
+            scope=SCOPE,
+            code_challenge=self.code_challenge,
+            code_challenge_method="S256",
+            **static_params,
+        )
+        click.echo(
+            "Complete the login via GitGuardian client. Launching browser to:\n"
```

Comment:
I think "client" is not the right word here. I would say something like this:

> To complete the login process, follow the instructions from {request_uri}.
> Opening your web browser now...

Reply:
I copied this one from Vault auth client, I can change it

Reply:
Ah right. Vault uses "your OIDC provider" instead of "GitGuardian client", but I don't think it's super clear either. I'd prefer if we changed it.